JASP-Engine/JASP/R/commonerrorcheck.R

```diff
@@ -617,4 +620,107 @@
   return(result)
 }
 
+.checkModelInteractions <- function(modelTerms) {
+  # In case of interactions, check whether all main effects and lower-order interaction terms are in the model
+  # Args:
+  #   modelTerms: a list of models terms, generated by specifying `listViewType: "Interaction"` in an AssignedVariablesList
+
+  result <- list(error=FALSE, errorVars=NULL)
+
+  if (length(modelTerms) == 0)
+    return(result)
+
+  for (term in modelTerms) {
+
+    if (is.null(names(term)) || !"components" %in% names(term))
+      stop("the modelterms list must contain the named item `components`")
+      
+    components <- term$components
+    if (length(components) < 2)
+      next
+
+    numChildren <- 2^length(components) - 1
+    inclusion <- sapply(modelTerms, 
+      function(termToFind) {
+          componentsToFind <- termToFind$components
+          if (sum(componentsToFind %in% components) == length(componentsToFind))
+            return(TRUE)
+          return(FALSE)
+      }
+    )
+
+    if (sum(inclusion) != numChildren)
+      result$error <- TRUE
+
+  }
 
+  return(result)
+}
+
+.checkNegativeValues <- function(dataset, target, grouping=NULL, groupingLevel=NULL) {
+  # Check for negative values in the dataset. 
+  # Args:
+  #   dataset: JASP dataset.
+  #   target: String vector indicating the target variables.
+  #   grouping: String vector indicating the grouping variables.
+  #   groupingLevel: Vector indicating the level of each of the grouping variables.
+  
+  result <- list(error=FALSE, errorVars=NULL)
+  
+  findNegativeValues <- function(x) {
+    return(any(x < 0))
+  }
+  
+  for (v in target) {
+    
+    if (is.factor(dataset[[.v(v)]])) { # Coerce factor to numeric.
+      dataset[[.v(v)]] <- as.numeric(as.character(dataset[[.v(v)]]))
+    } 
+    
+    if (length(grouping) > 0 && length(groupingLevel) > 0) { 
+      hasNegativeValues <- .applyOnGroups(findNegativeValues, dataset, v, grouping, groupingLevel)
+    } else {
+      hasNegativeValues <- findNegativeValues(dataset[[.v(v)]])
+    }
+    
+    if (hasNegativeValues) {
+      result$error <- TRUE
+      result$errorVars <- c(result$errorVars, v)
+    }
+    
+  }
+  
+  return(result)
+}
+
+.checkMissingValues <- function(dataset, target, grouping=NULL, groupingLevel=NULL) {
+  # Check for missing values in the dataset. 
+  # Args:
+  #   dataset: JASP dataset.
+  #   target: String vector indicating the target variables.
+  #   grouping: String vector indicating the grouping variables.
+  #   groupingLevel: Vector indicating the level of each of the grouping variables.
+  
+  result <- list(error=FALSE, errorVars=NULL)
+  
+  findMissingValues <- function(x) {
+    return(any(is.na(x)))
```

Comment:
`anyNA` should do the trick

Reply:
Love your knowledge of these niche R functions